zeppelin-web-angular/e2e/tests/notebook/keyboard/notebook-keyboard-shortcuts.spec.ts

```diff
@@ -0,0 +1,867 @@
+    test('should cut line with Control+K', async () => {
+      // Given: Code editor with content
+      await keyboardPage.focusCodeEditor();
+      await keyboardPage.setCodeEditorContent('first line\nsecond line\nthird line');
+
+      const initialContent = await keyboardPage.getCodeEditorContent();
+      expect(initialContent).toContain('first line');
+
+      // Additional wait and focus for Firefox compatibility
+      const browserName = test.info().project.name;
+      if (browserName === 'firefox') {
+        await keyboardPage.page.waitForTimeout(200);
+        // Ensure Monaco editor is properly focused
+        const editorTextarea = keyboardPage.page.locator('.monaco-editor textarea').first();
+        await editorTextarea.click();
+        await editorTextarea.focus();
+        await keyboardPage.page.waitForTimeout(200);
+      }
+
+      // When: User presses Control+K (cut to end of line)
+      await keyboardPage.pressCutLine();
+
+      // Then: First line content should be cut (cut from cursor position to end of line)
+      await keyboardPage.page.waitForTimeout(500);
+      const finalContent = await keyboardPage.getCodeEditorContent();
+      expect(finalContent).toBeDefined();
+      expect(typeof finalContent).toBe('string');
+
+      // Verify the first line was actually cut
+      expect(finalContent).toContain('first line');
+      expect(finalContent).toContain('second line');
+      expect(finalContent).not.toContain('third line');
+    });
+  });
+
+  test.describe('ParagraphActions.PasteLine: Control+Y', () => {
+    test('should paste line with Control+Y', async () => {
+      // Given: Content in the editor
+      await keyboardPage.focusCodeEditor();
+      await keyboardPage.setCodeEditorContent('line to cut and paste');
+
+      // Wait for content to be properly set before verifying
+      await keyboardPage.page.waitForTimeout(500);
+
+      const initialContent = await keyboardPage.getCodeEditorContent();
+      // Debug: Log the actual content and its character codes
+      console.log('Initial content:', JSON.stringify(initialContent));
+      console.log('Expected:', JSON.stringify('line to cut and paste'));
+
+      // Use a more robust assertion that handles encoding issues
+      const expectedText = 'line to cut and paste';
+      expect(
+        initialContent.includes(expectedText) ||
+          initialContent.normalize().includes(expectedText) ||
+          initialContent.replace(/\s+/g, ' ').trim().includes(expectedText)
+      ).toBeTruthy();
+
+      // Cut the line first
+      await keyboardPage.pressCutLine();
+      await keyboardPage.page.waitForTimeout(500);
+
+      // When: User presses Control+Y to paste
+      await keyboardPage.pressPasteLine();
+
+      // Then: Content should be pasted back
+      await keyboardPage.page.waitForTimeout(500);
+      const finalContent = await keyboardPage.getCodeEditorContent();
+      expect(finalContent.length).toBeGreaterThan(0);
```

Comment:
Could we also assert that the content actually changes?
For example, after `Cut Line` we check that the editor content is reduced (something disappears), then clear the input and verify it’s empty, and finally after `Paste Line` assert that the original content is restored.